### DISCUSSION TITLE
How to stop training after a fixed number of steps?

### DISCUSSION THREAD

**Discussion by CloudyDory** [thu-ml/tianshou#845] [Q&A]
The default OnPolicyTrainer has an option to stop training after a fixed number of episodes. However, sometimes it is desirable to stop after a fixed number of total environment steps that the agent has experienced with. How to implement it in Tianshou?

**Comment by Trinkle23897** [comment]
https://github.com/thu-ml/tianshou/blob/4ac407c78f58102fa7f38ded6bfc1e42c703a4a7/tianshou/trainer/onpolicy.py#L28-L35

You can specify either `step_per_collect` or `episode_per_collect`

**Reply by CloudyDory** [reply]
Hi, I do not mean to specify the number of steps before a network update. I hope to specify the total number of environment steps in an entire training session. For example, stop training after 3.0e7 steps.

**Reply by CloudyDory** [reply]
I hope to do so because when training Atari games, we usually plot the epsiodic return as a function of total number of environment steps that the agent has experienced. And I would like to specify the maximum number of environment steps, not the maximum number of episodes.

**Reply by Trinkle23897** [reply]
https://github.com/thu-ml/tianshou/blob/master/examples/atari/atari_ppo.py

For example, if you want to set total steps 3e7, you can set epoch=100 with --step-per-epoch=3e5: `total_steps == epoch * step_per_epoch`; you can change number of epoch to control the evaluation interval

**Reply by CloudyDory** [reply]
Thank you! Another small question: what will happen if the game does not end when reaching the maximum number of steps in an epoch? Will the game stop immediately, or will it stop only after the current episode ends (so the actual steps will be slightly larger than the maximum steps)?

**Reply by Trinkle23897** [reply]
it will stop stepping env, switch to test env, do one evaluation, then continue stepping the partially stepped training env
